### ISSUE TITLE
Slow TFT Update

### ISSUE THREAD

**Issue by fencer** [Blackymas/NSPanel_HA_Blueprint#212]
I have setup my NSPanel an started to upload the TFT file however it has been over 8 hours so far only 38%. It's not stuck, just slowly incrementing. Is this normal?

**Comment by bkbartk** [comment]
I have about the same issue,
2 hours 12%
the log basically says this
<html><body>


18:35:49 | [E] | [uart:015] | Reading from UART timed out at byte 0!
-- | -- | -- | --
18:35:49 | [W] | [nextion:078] | Nextion is not connected!
18:35:50 | [D] | [sensor:127] |  


</body>
</html>

and during this nextion update the switches aren't working.

On my old firmware an update like this took about 20minutes which I already thought was long

**Comment by Blackymas** [comment]
Hi,

Slow donwload has nothing to do with the code and seems to be a network/wifi or internet problem.

But you could just download the TFT and upload it to HA in the www folder and then use our built-in service.

http://192.168.4.250:8123/local/nspanel/nspanel.tft

![Bildschirmfoto 2023-01-21 um 20 06 48](https://user-images.githubusercontent.com/41958506/213883113-a51b31dc-6e11-4106-bc5d-98e3bd93dcc3.png)

**Comment by bkbartk** [comment]
for cancelling the current action you need to restart the HMI.

So I quickly added a button for that.

```
  - platform: gpio
    name: restart Nextion
    id: restartNextion
    pin:
      number: 4
    on_turn_on:
      - script.execute: failsafe
    on_turn_off:
      - script.stop: failsafe
 ```
```
script:
  - id: failsafe
    then:
      - delay: 3s
      - switch.turn_off: restartNextion
```

So I'm able to reboot manually

I didn't use HA to store the file because HA only is accessible using https. but on my local pc.

The log now says this
```
20:42:56 | [D] | [number:113] | New number value: 10.000000
-- | -- | -- | --
20:42:56 | [D] | [number:012] | 'lamp-kamer-voor Display Brightness Dimdown': Sending state 10.000000
20:42:56 | [D] | [sensor:127] | 'ntc_source': Sending state 1.89200 V with 2 decimals of accuracy
20:42:56 | [D] | [resistance:039] | 'resistance_sensor' - Resistance 15050.0Ω
20:42:56 | [D] | [sensor:127] | 'resistance_sensor': Sending state 15050.00000 Ω with 1 decimals of accuracy
20:42:56 | [D] | [sensor:127] | 'resistance_sensor': Sending state 15050.00000 Ω with 1 decimals of accuracy
20:42:56 | [D] | [ntc:026] | 'lamp-kamer-voor Temperature' - Temperature: 16.1°C
20:42:56 | [D] | [sensor:127] | 'lamp-kamer-voor Temperature': Sending state 16.07562 °C with 1 decimals of accuracy
20:42:56 | [D] | [switch:056] | 'lamp-kamer-voor Nextion inited': Sending state ON
20:43:00 | [D] | [sensor:127] | 'lamp-kamer-voor RSSI': Sending state -39.00000 dBm with 0 decimals of accuracy
20:43:05 | [D] | [nextion_upload:169] | Connected
20:43:05 | [D] | [nextion_upload:175] | Requesting URL: http://192.168.180.148/nspanel.tft
20:43:05 | [D] | [nextion_upload:209] | Updating Nextion NX4832F035_011C...
20:43:05 | [D] | [nextion_upload:235] | Waiting for upgrade response
20:43:05 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:05 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:05 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:05 | [D] | [nextion_upload:239] | Upgrade response is  4
20:43:05 | [D] | [nextion_upload:242] | Available 0 : 0x00
20:43:05 | [D] | [nextion_upload:242] | Available 1 : 0x00
20:43:05 | [D] | [nextion_upload:242] | Available 2 : 0x00
20:43:05 | [D] | [nextion_upload:242] | Available 3 : 0x05
20:43:05 | [D] | [nextion_upload:246] | preparation for tft update done
20:43:05 | [D] | [nextion_upload:283] | Allocating buffer size 65536, Heap size is 164580
20:43:05 | [D] | [nextion_upload:305] | Updating tft from "http://192.168.180.148/nspanel.tft" with a file size of 7346972 using 65536 chunksize, Heap Size 97224
20:43:05 | [D] | [nextion_upload:044] | Requesting range: bytes=0-16383
20:43:07 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:07 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:07 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:07 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:07 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:07 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:09 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:09 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:09 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:09 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:09 | [E] | [uart:015] | Reading from UART timed out at byte 0!
20:43:09 | [D] | [nextion_upload:117] | Nextion reported new range 134744072
20:43:09 | [D] | [nextion_upload:316] | Heap Size 105348, Bytes left -127397100
20:43:09 | [D] | [nextion_upload:318] | Successfully updated Nextion!
20:43:09 | [D] | [nextion_upload:324] | Restarting Nextion

```
but after 15 minutes it's still at 0%
(time is incorrect, I restarted again to copy the logging)

I think there is some issue with the uart, not sure.
I think I will install tasmota, then flash and then go back to esphome( I did this also to get nspanel_blank.tft working)

**Comment by fencer** [comment]
highly doubt this is a signal problem (12hrs is to download 50% of a 7mb file would mean I am below 1 kb/s :/ )
![Screenshot_20230121-204100](https://user-images.githubusercontent.com/620966/213884623-9043d72f-34a8-4c3a-800f-5b22eff802c7.png)


> Hi,
> 
> Slow donwload has nothing to do with the code and seems to be a network/wifi or internet problem.
> 
> But you could just download the TFT and upload it to HA in the www folder and then use our built-in service.
> 
> http://192.168.4.250:8123/local/nspanel/nspanel.tft
> 
> ![Bildschirmfoto 2023-01-21 um 20 06 48](https://user-images.githubusercontent.com/41958506/213883113-a51b31dc-6e11-4106-bc5d-98e3bd93dcc3.png)

**Comment by bkbartk** [comment]
ok,
I Installed tasmota, flashed the tft file
this took 205 seconds according to the log,
then I installed esphome again,
now I'm at the blue"Please wait" screen

So I don't know what it is, but I don't think it's a network issue.

**Comment by danwright87** [comment]
The upload of TFT via ESP Home is slow in general as its over serial and is downloaded in small chunks, with the additional latency of doing that over the internet the the delays become substantial. Outside of this project ive had tft upload from a remote source take 5+ hours down to maybe 10 minutes from a local web server, this is not a issue with with project but a ESP Home trait. Using local source as recommended is the best action if you are struggling with how long upload takes. 

Tasmota does seem to handle it much better and has been able to rescue bricked nextion displays for me that ESP Home wouldn't connect to.

**Comment by Blackymas** [comment]
> The upload of TFT via ESP Home is slow in general as its over serial and is downloaded in small chunks, with the additional latency of doing that over the internet the the delays become substantial.

right and that's what I meant - many have no problem with it I myself use 7 NSPanel without problems.

But since we have seen that some have problems with it, we have built extra the service for an easy and fast local upload :)

**Comment by fencer** [comment]
> for cancelling the current action you need to restart the HMI.
> 
> So I quickly added a button for that.
> 
> ```
>   - platform: gpio
>     name: restart Nextion
>     id: restartNextion
>     pin:
>       number: 4
>     on_turn_on:
>       - script.execute: failsafe
>     on_turn_off:
>       - script.stop: failsafe
> ```
> 
> ```
> script:
>   - id: failsafe
>     then:
>       - delay: 3s
>       - switch.turn_off: restartNextion
> ```
> 
> So I'm able to reboot manually
> 
> I didn't use HA to store the file because HA only is accessible using https. but on my local pc.
> 
> The log now says this
> 
> ```
> 20:42:56 | [D] | [number:113] | New number value: 10.000000
> -- | -- | -- | --
> 20:42:56 | [D] | [number:012] | 'lamp-kamer-voor Display Brightness Dimdown': Sending state 10.000000
> 20:42:56 | [D] | [sensor:127] | 'ntc_source': Sending state 1.89200 V with 2 decimals of accuracy
> 20:42:56 | [D] | [resistance:039] | 'resistance_sensor' - Resistance 15050.0Ω
> 20:42:56 | [D] | [sensor:127] | 'resistance_sensor': Sending state 15050.00000 Ω with 1 decimals of accuracy
> 20:42:56 | [D] | [sensor:127] | 'resistance_sensor': Sending state 15050.00000 Ω with 1 decimals of accuracy
> 20:42:56 | [D] | [ntc:026] | 'lamp-kamer-voor Temperature' - Temperature: 16.1°C
> 20:42:56 | [D] | [sensor:127] | 'lamp-kamer-voor Temperature': Sending state 16.07562 °C with 1 decimals of accuracy
> 20:42:56 | [D] | [switch:056] | 'lamp-kamer-voor Nextion inited': Sending state ON
> 20:43:00 | [D] | [sensor:127] | 'lamp-kamer-voor RSSI': Sending state -39.00000 dBm with 0 decimals of accuracy
> 20:43:05 | [D] | [nextion_upload:169] | Connected
> 20:43:05 | [D] | [nextion_upload:175] | Requesting URL: http://192.168.180.148/nspanel.tft
> 20:43:05 | [D] | [nextion_upload:209] | Updating Nextion NX4832F035_011C...
> 20:43:05 | [D] | [nextion_upload:235] | Waiting for upgrade response
> 20:43:05 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:05 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:05 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:05 | [D] | [nextion_upload:239] | Upgrade response is  4
> 20:43:05 | [D] | [nextion_upload:242] | Available 0 : 0x00
> 20:43:05 | [D] | [nextion_upload:242] | Available 1 : 0x00
> 20:43:05 | [D] | [nextion_upload:242] | Available 2 : 0x00
> 20:43:05 | [D] | [nextion_upload:242] | Available 3 : 0x05
> 20:43:05 | [D] | [nextion_upload:246] | preparation for tft update done
> 20:43:05 | [D] | [nextion_upload:283] | Allocating buffer size 65536, Heap size is 164580
> 20:43:05 | [D] | [nextion_upload:305] | Updating tft from "http://192.168.180.148/nspanel.tft" with a file size of 7346972 using 65536 chunksize, Heap Size 97224
> 20:43:05 | [D] | [nextion_upload:044] | Requesting range: bytes=0-16383
> 20:43:07 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:07 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:07 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:07 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:07 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:07 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:08 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:09 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:09 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:09 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:09 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:09 | [E] | [uart:015] | Reading from UART timed out at byte 0!
> 20:43:09 | [D] | [nextion_upload:117] | Nextion reported new range 134744072
> 20:43:09 | [D] | [nextion_upload:316] | Heap Size 105348, Bytes left -127397100
> 20:43:09 | [D] | [nextion_upload:318] | Successfully updated Nextion!
> 20:43:09 | [D] | [nextion_upload:324] | Restarting Nextion
> ```
> 
> but after 15 minutes it's still at 0% (time is incorrect, I restarted again to copy the logging)
> 
> I think there is some issue with the uart, not sure. I think I will install tasmota, then flash and then go back to esphome( I did this also to get nspanel_blank.tft working)

so did you reboot the nextion display mid-flash?

**Comment by bkbartk** [comment]
Yes. You get a “system data error” in the screen. But you can start the flash again.

**Comment by fencer** [comment]
I got the "system data error" but not able to re-flash. After trying Tasmota and using the nextion editor it's self without any joy I was able to flash the blank using the this tool. Tried to flash the TFT again but now I'm back stuck with a blank white screen and unable to flash again (not event the blank). 

`
11:34:38 | [D] | [nextion_upload:235] | Waiting for upgrade response
-- | -- | -- | --
11:34:38 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:38 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:38 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:38 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:38 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:38 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:39 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:39 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:39 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:39 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:39 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:39 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:39 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:39 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:39 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:39 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:40 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:40 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:40 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:40 | [E] | [uart:015] | Reading from UART timed out at byte 0!
11:34:40 | [D] | [nextion_upload:239] | Upgrade response is  20
11:34:40 | [D] | [nextion_upload:242] | Available 0 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 1 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 2 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 2 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 3 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 4 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 5 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 6 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 7 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 8 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 9 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 9 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 10 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 10 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 11 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 12 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 13 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 13 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 14 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 15 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 16 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 17 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 18 : 0x00
11:34:40 | [D] | [nextion_upload:242] | Available 19 : 0x00
11:34:40 | [D] | [nextion_upload:248] | preparation for tft update failed 0 ""
11:34:40 | [D] | [nextion_upload:324] | Restarting Nextion
11:34:40 | [D] | [nextion_upload:324] | Restarting Nextion
11:34:41 | [D] | [nextion_upload:327] | Restarting esphome`

This happens to be the same error I kept getting when I had the “System data ERRROR!” on screen

**Comment by mic267** [comment]
Ich hatte auch das selbe Problem. Ich glaube alle die das Nextion NX4832F035_011C Display verbaut haben müssen zuerst
das blank.tft flashen.

**Comment by mic267** [comment]
Nach dem flashen in Ha update tft. dann nochmal flashen mit der nspanel.tft

**Comment by fencer** [comment]
Success! I changed the baud rate from 9600 back to 115200 and disabled debug logging. Flashed the display and then I was able to flash the TFT.

**Comment by bkbartk** [comment]
It you want to flash using tasmota
`FlashNextion http://192.168.180.148/nspanel.tft`
it's important to have the file available on http, not https, however the download on tasmota is faster it does not support https fully.

@fencer 
thanks, I actually think this baud rate thing is key here, when I need to update again, I will check how this works out.

**Comment by fencer** [comment]
@bkbartk I messed about with the baud rate (that's how I think I got the blank screen to work). I'm currently trying to understand why the display now is stuck at initializing with the HA background

**Comment by bkbartk** [comment]
for the HA background,
you need to install the blueprint.
it says the name should be equal, but I needed to replace "-" with "_"
and I changed the name of the buttons and relays in esphome,
that I needed to undo.

After that I changed the buttons a bit so the toggle action is handled in esphome instead of HA. Buttons should also work if wifi or ha is down in my opinion.

**Comment by fencer** [comment]
> it says the name should be equal, but I needed to replace "-" with "_"

That did the trick! Thanks!!!

**Comment by iz3man** [comment]
> http://192.168.4.250:8123/local/nspanel/nspanel.tft

Where did you set the baudrate?

**Comment by Blackymas** [comment]
> Where did you set the baudrate?
> 
> I'm getting these messages in the ESPhome log:
> 
> [17:27:07][W][nextion:078]: Nextion is not connected! [17:27:07][E][uart:015]: Reading from UART timed out at byte 0! [17:27:07][W][nextion:078]: Nextion is not connected! [17:27:08][E][uart:015]: Reading from UART timed out at byte 0! [17:27:08][W][nextion:078]: Nextion is not connected! [17:27:08][E][uart:015]: Reading from UART timed out at byte 0! [17:27:08][W][nextion:078]: Nextion is not connected! [17:27:09][E][uart:015]: Reading from UART timed out at byte 0! [17:27:09][W][nextion:078]: Nextion is not connected! [17:27:10][E][uart:015]: Reading from UART timed out at byte 0! [17:27:10][W][nextion:078]: Nextion is not connected! [17:27:10][E][uart:015]: Reading from UART timed out at byte 0! [17:27:10][W][nextion:078]: Nextion is not connected!

Es ist nicht nötig die Baudrate zu ändern.

Du musst eigentlich nur die nspanel_blank.tft herunterladen und in Ha unter www hochladen.

Danach den eingebauten Service benutzen um das tft lokal auf dein TFT zu flashen.


https://github.com/Blackymas/NSPanel_HA_Blueprint/issues/166
https://github.com/Blackymas/NSPanel_HA_Blueprint/issues/177
https://github.com/Blackymas/NSPanel_HA_Blueprint/issues/197

**Comment by iz3man** [comment]
Thanks. #197 worked. Installing blank file, then do the normal upload! :)

**Comment by geekyclarkey** [comment]
I spent days tinkering with this. my first aft update took 24 hours. I think it was something to do with the baud rate. After flashing with Tasmota and then back to esp home I got the baud rate to work using 115200. I think as 9600 was causing it to update slow. I then flashed the nspanel_blank.tft file (in the custom_configuration folder) then the normal one after. I also had a "-" in the esp name so replacing that with a "_" got the whole thing working as it should.
